### DISCUSSION TITLE
Recomended way to check the status of stream(s)

### DISCUSSION THREAD

**Discussion by avmarko** [oliver-zehentleitner/unicorn-binance-websocket-api#337] [Q&A] [answered]
Hi. I am experimenting on a script based on the "stream_everything" example. It creates multiple streams, with multiple-channel/market subscriptions. I 'm also using a custom stream-data-processor based on the "process_streams" example. What is the best practice to check that 1) a specific stream is actually created and ready 2) all streams are ready. Same need goes when I subscribe/unsubscribe to addition markets.  I guess the answer relies on all these {"result":null,"id":1} messages, but I don't know how to distinct which stream-id is each one of the for. I wonder if there is a recommended and easy approach to this...

**Comment by oliver-zehentleitner** [comment]
Hello!

Basically:
Once a stream has been created and worked, there are only 2 reasons why it might not work again:
1. for a private API stream, API key/secret pair have become invalid.
2. no network. This is usually a temporary problem and the stream will reconnect (at least that's how it's supposed to) as soon as the host is back online.

There are ways to [check or query the current status](https://unicorn-binance-websocket-api.docs.lucit.tech/unicorn_binance_websocket_api.html#unicorn_binance_websocket_api.manager.BinanceWebSocketApiManager.get_stream_info) and you can be informed about status changes with [stream signals - as buffer or callback](https://github.com/LUCIT-Systems-and-Development/unicorn-binance-websocket-api/wiki/%60stream_signal_buffer%60).

I have put all the info you need in the [documentation](https://unicorn-binance-websocket-api.docs.lucit.tech/). Use the docs internal search function and have a look at the topics, monitoring, get_results_from_endpoints(), get_results_from_endpoints() and stream_labels. 

If you go through the readme once from top to bottom, it will present and link to everything relevant from UBWA.

I plan to release a Medium article and YouTube video on this topic soon (mastering ubwa). You can follow me there, then you will be informed about the publication. First the article will appear, then the video.
https://medium.com/@oliverzehentleitner
https://www.youtube.com/channel/UCp57u8vaPqEWQTlfpYn-KnQ

**Comment by avmarko** [comment]
@oliver-zehentleitner thank you for your answer. Already subscribed and looking forward for your Medium article.

BTW and since you mentioned "once a stream has been created and worked":  
How can I be notified about that exact moment when the stream started and is working? If I wait for the first incoming data-message from the stream to confirm subscription, then it may take some time (imagine e.g. a "trades" stream for a market-pair with rare transactions). The reason I am searching for the recommended way to do this, is that the python script will act as a binance-gateway to an external system - and this system needs to be notified when the web-socket monitoring is actually active. I am using MSMQ messaging for this, so I need to know when to send a confirmation message. The message will be something like: "OK - Market X subscribed to stream Y". Or "OK - all requested markets subscribed to all requested streams").

**Comment by oliver-zehentleitner** [comment, marked as answer]
there are different approaches here.

Firstly, you can activate exceptions if a stream cannot start. then you have the stream signals. The "result" message should also be a good indication. you could also query the subscriptions with get_stream_subscriptions() after wait_till_stream_has_started() returns true.

A mapping from "subscriptions and confirmed subscriptions" would be nice, but is not implemented.

**Reply by avmarko** [reply]
Thanks @oliver-zehentleitner :-) 
I will experiment with these suggestions
